cudagraph trees

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #96775
* __->__ #89146
* #96384

CUDA Graph Trees 

Design doc: https://docs.google.com/document/d/1ZrxLGWz7T45MSX6gPsL6Ln4t0eZCSfWewtJ_qLd_D0E/edit

Not currently implemented :

- Right now, we are using weak tensor refs from outputs to check if a tensor has dies. This doesn't work because a) aliasing, and b) aot_autograd detaches tensors (see note [Detaching saved tensors in AOTAutograd]). Would need either https://github.com/pytorch/pytorch/issues/91395 to land to use storage weak refs or manually add a deleter fn that does what I want. This is doable but theres some interactions with the caching allocator checkpointing so saving for a stacked pr.

- Reclaiming memory from the inputs during model recording. This isn't terribly difficult but deferring to another PR. You would need to write over the input memory during warmup, and therefore copy the inputs to cpu. Saving for a stacked pr.

- Warning on overwriting previous generation outputs. and handling nested torch.compile() calls in generation tracking

Differential Revision: [D43999887](https://our.internmc.facebook.com/intern/diff/D43999887)

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

eellison committed:
cudagraph tapes'

[ghstack-poisoned]

ezyang commented:
What's the inputs (?) You want inputs in the pool. In conventional usage, we'll be copying in the training data to the cuda graph to kick off the run.

ezyang commented:
> To avoid overwriting the live output values of a previous invocation, we can either check that the backward has been run for a cudagraph forward (not clear to me this is sufficient ?). We potentially could also do weak-ref ref counting.

For correctness, don't you have to do the "wait until all the incoming refcounts are dead". Doesn't have to be weak-ref, IIRC?

ezyang commented on torch/_inductor/compile_fx.py:
```diff
@@ -219,88 +224,630 @@ def remove_unaligned_input_idxs(inputs, static_input_idxs):
 
 
-def cudagraphify_impl(model, inputs, static_input_idxs=()):
+@dataclasses.dataclass(frozen=True)
```
Bah, does this really have to live in torch._inductor? This should work for any backend right?

eellison replied:
Yea, can move out in the next pr

ezyang replied:
In fact, the CUDAGraphTape isn't even a compiler only abstraction; the eager mode cuda graphs api could hypothetically use this too.

eellison replied:
You would need to add the same guards that exist in dynamo, but otherwise, yes.

ezyang commented on torch/_inductor/compile_fx.py:
```diff
@@ -219,88 +224,630 @@ def remove_unaligned_input_idxs(inputs, static_input_idxs):
+    to follow the same order and have the same output tensor livespans. To remove
+    unnecessary coupling of cuda graphs (and additional imposed invariants),
+    the tape manager will end a currently recording tape whenever it is valid  - when
+    the memory pool no longer has any live allocations.
```
This is bad for perf no? Once you finish a cuda graph tape, the memory pool associated with the tape is permanently inaccessible. Even if the pool has no more live allocations, if you continue on the tape you can reuse this memory.

As for what the right way to detect if the tape is done or not, I would probably have suggested snipping it when you "revisit" a node that you've already seen before.

eellison replied:
All of the tapes use the same memory pool. Will add comment to make that clear.

ezyang replied:
I feel like there should be more invariants on the tapes. Can a subgraph be a member of multiple tapes?

eellison replied:
Not right now. If you run 

`foo -> foo2` that will be a single recording. On the next execution, if you ran `foo -> foo3`, the caching allocator for the memory pool would not represent the current state of allocations, so you not trivially trace `foo3` then with the same memory pool. 

You would need to make the cuda caching allocator resumable by snapshotting its state at the end of the foo recording, resetting it, and then record `foo3`. Was planning on doing this in next pr on the stack (and won't merge until they are all ready).

ezyang replied:
sounds complicated lol

eellison commented on torch/_inductor/compile_fx.py:
```diff
@@ -24,6 +27,8 @@
 )
 from .virtualized import V
 
+WeakRef = Any
```
`weakref.ref[Type]` gives runtime error :/ not sure of a better way to do it

ezyang replied:
it looks like you might need https://mypy.readthedocs.io/en/latest/runtime_troubles.html#future-annotations-import-pep-563

eellison reviewed (commented):
> What's the inputs (?) You want inputs in the pool. In conventional usage, we'll be copying in the training data to the cuda graph to kick off the run.

the inputs are being copied into the pool

eellison commented:
> For correctness, don't you have to do the "wait until all the incoming refcounts are dead". Doesn't have to be weak-ref, IIRC?

Uh, this is a bit outdated, check PR.

eellison committed:
Update on "cudagraph tapes'"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

This is a first PR implementing the following plan of POR:

- We will use a single Memory Pool across all cudagraphs. The memory pool will contain inputs (?), intermediates, and outputs. CUDAGraphs memory pool ensures that memory allocated within graph capture is not allocated outside of the memory pool, so that memory addresses remain static during graph replay. However, within the memory pool, normal storage allocation occurs, so there is no memory overhead. So long as we invoke separate CUDAGraphs in a consistent order as they are recorded, and do not retain extraneous references to Tensors, there should be no overhead. We will keep a global tape of the initial graph recordings and subsequent usages. Enforcing consistent CUDAGraph usage will also allow us to avoid copying over inputs to a CUDAGraph that are the outputs of a prior one. 
- To avoid “static outputs” being permanently allocated, which interferes with both memory being reclaimed within CUDAGraphs and [gradient stealing](https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/functions/accumulate_grad.h#L52), we will reconstruct output tensors Tensors with the appropriate fixed storage data pointers and other metadata on each cudagraph invocation. 
- When a CUDAGraph usage diverges from the existing tape we can 1. free the memory pool to the rest of the cuda allocator, invalidating the CUDAGraphs of our tape but keeping the existing Tensors alive or 2. Run torch inductor without cudagraphs for the remainder of the tape. The latter would have some memory penalty because it would not use memory from the memory pool 
- Because cudagraphs freeze memory addresses, the invocation of a cudagraph will replace whatever values the memory had previously. To avoid overwriting the live output values of a previous invocation, we can either check that the backward has been run for a cudagraph forward (not clear to me this is sufficient ?). We potentially could also do weak-ref ref counting.

Not everything is implemented yet, and I will not land this pr without subsequent follow ups. 

Follows ups, for correctness/minimal viable product in terms of priority include :
- Registering the cudagraphs as thread local state that needs to be copied over in autograd
- Ensuring clean up works successfully
- Adding tests for all the edge cases provided reviewers are okay with approach
- Handling subsequent usage when a graph invariant is broken, i.e. moving cuda graph allocations from the private memory pool to eager

And then further work would be:
- Allowing multiple, separate cudagraph paths by making the cuda caching allocator resumable and doing new recordings. 
- Adding configs to weaken some of the correctness conditions (for power user who can verify correctness)



[ghstack-poisoned]

ezyang commented on torch/_inductor/compile_fx.py:
```diff
@@ -219,88 +224,630 @@ def remove_unaligned_input_idxs(inputs, static_input_idxs):
+
+        self.debug_fail_counter = 0
+
+    def increment_recording_tape(self) -> GraphID:
```
would be nice to have docblocks/comments for these methods

ezyang commented on torch/_inductor/compile_fx.py:
```diff
@@ -219,88 +224,630 @@ def remove_unaligned_input_idxs(inputs, static_input_idxs):
 
+    def __init__(self, tape_manager):
+
+        self.tape_manager = tape_manager
```
This looks like reference cycle

ezyang commented on test/inductor/test_torchinductor.py:
```diff
@@ -4924,6 +4924,87 @@ def fn(a: torch.Tensor) -> torch.Tensor:
 
+    class CudaGraphTests(TestCase):
+        def tearDown(self):
+            torch._inductor.compile_fx.tape_manager_container.tape_manager = None
```
can we avoid having global state (or, have a non-global state api that the unit tests can use?)

eellison replied:
Talked with jason abt this a bit. What makes this difficult is that the compiled function is installed on the bytecode. So if you have something like: 

```
import torch
import torch._dynamo
import torch._inductor
import refcycle
import weakref

foo_dies = False

def test_closure():
    @torch._dynamo.optimize()
    def foo(x):
        return x + 1 + 2

    def finalize():
        global foo_dies
        foo_dies = True

    weakref.finalize(foo, finalize)

    return foo(torch.rand([4], device="cuda"))

out = test_closure()
snapshot = refcycle.snapshot()
```

While the `foo` will get deallocated, the compiled function that is installed on foo's bytecode will not. The second time you invoke `test_closure`, the function is not recompiled.

Could we call `torch._dynamo.reset` before after invoking all pytorch tests ?

eellison replied:
I added a test where we are showing the correct behavior wrt/ `reset()`.

ezyang replied:
I once tried calling torch._dynamo.reset() before all dynamo unit tests and it made the test suite run irreparably slowly. But yes, resetting seems like the right thing to do for the *dynamo* side integration... but pursuant to my point about this being a valid eager mode only interface, I think you should probably consider doing a bunch of tests without dynamo at all.

eellison committed:
Update on "cudagraph tapes'"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

This is a first PR implementing the following plan of POR:

- We will use a single Memory Pool across all cudagraphs. The memory pool will contain inputs (?), intermediates, and outputs. CUDAGraphs memory pool ensures that memory allocated within graph capture is not allocated outside of the memory pool, so that memory addresses remain static during graph replay. However, within the memory pool, normal storage allocation occurs, so there is no memory overhead. So long as we invoke separate CUDAGraphs in a consistent order as they are recorded, and do not retain extraneous references to Tensors, there should be no overhead. We will keep a global tape of the initial graph recordings and subsequent usages. Enforcing consistent CUDAGraph usage will also allow us to avoid copying over inputs to a CUDAGraph that are the outputs of a prior one. 
- To avoid “static outputs” being permanently allocated, which interferes with both memory being reclaimed within CUDAGraphs and [gradient stealing](https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/functions/accumulate_grad.h#L52), we will reconstruct output tensors Tensors with the appropriate fixed storage data pointers and other metadata on each cudagraph invocation. 
- When a CUDAGraph usage diverges from the existing tape we can 1. free the memory pool to the rest of the cuda allocator, invalidating the CUDAGraphs of our tape but keeping the existing Tensors alive or 2. Run torch inductor without cudagraphs for the remainder of the tape. The latter would have some memory penalty because it would not use memory from the memory pool 
- Because cudagraphs freeze memory addresses, the invocation of a cudagraph will replace whatever values the memory had previously. To avoid overwriting the live output values of a previous invocation, we can either check that the backward has been run for a cudagraph forward (not clear to me this is sufficient ?). We potentially could also do weak-ref ref counting.

Not everything is implemented yet, and I will not land this pr without subsequent follow ups. 

Follows ups, for correctness/minimal viable product in terms of priority include :
- Registering the cudagraphs as thread local state that needs to be copied over in autograd
- Ensuring clean up works successfully
- Adding tests for all the edge cases provided reviewers are okay with approach
- Handling subsequent usage when a graph invariant is broken, i.e. moving cuda graph allocations from the private memory pool to eager

And then further work would be:
- Allowing multiple, separate cudagraph paths by making the cuda caching allocator resumable and doing new recordings. 
- Adding configs to weaken some of the correctness conditions (for power user who can verify correctness)



[ghstack-poisoned]

eellison committed:
Update on "cudagraph tapes'"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

This is a first PR implementing the following plan of POR:

- We will use a single Memory Pool across all cudagraphs. The memory pool will contain inputs (?), intermediates, and outputs. CUDAGraphs memory pool ensures that memory allocated within graph capture is not allocated outside of the memory pool, so that memory addresses remain static during graph replay. However, within the memory pool, normal storage allocation occurs, so there is no memory overhead. So long as we invoke separate CUDAGraphs in a consistent order as they are recorded, and do not retain extraneous references to Tensors, there should be no overhead. We will keep a global tape of the initial graph recordings and subsequent usages. Enforcing consistent CUDAGraph usage will also allow us to avoid copying over inputs to a CUDAGraph that are the outputs of a prior one. 
- To avoid “static outputs” being permanently allocated, which interferes with both memory being reclaimed within CUDAGraphs and [gradient stealing](https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/functions/accumulate_grad.h#L52), we will reconstruct output tensors Tensors with the appropriate fixed storage data pointers and other metadata on each cudagraph invocation. 
- When a CUDAGraph usage diverges from the existing tape we can 1. free the memory pool to the rest of the cuda allocator, invalidating the CUDAGraphs of our tape but keeping the existing Tensors alive or 2. Run torch inductor without cudagraphs for the remainder of the tape. The latter would have some memory penalty because it would not use memory from the memory pool 
- Because cudagraphs freeze memory addresses, the invocation of a cudagraph will replace whatever values the memory had previously. To avoid overwriting the live output values of a previous invocation, we can either check that the backward has been run for a cudagraph forward (not clear to me this is sufficient ?). We potentially could also do weak-ref ref counting.

Not everything is implemented yet, and I will not land this pr without subsequent follow ups. 

Follows ups, for correctness/minimal viable product in terms of priority include :
- Registering the cudagraphs as thread local state that needs to be copied over in autograd
- Ensuring clean up works successfully
- Adding tests for all the edge cases provided reviewers are okay with approach
- Handling subsequent usage when a graph invariant is broken, i.e. moving cuda graph allocations from the private memory pool to eager

And then further work would be:
- Allowing multiple, separate cudagraph paths by making the cuda caching allocator resumable and doing new recordings. 
- Adding configs to weaken some of the correctness conditions (for power user who can verify correctness)



[ghstack-poisoned]

ezyang commented on torch/_inductor/compile_fx.py:
```diff
@@ -219,88 +225,649 @@ def remove_unaligned_input_idxs(inputs, static_input_idxs):
-            new_inputs.clear()
-            graph.replay()
-            return static_outputs
+def cudagraphify_impl(model, inputs, static_input_idxs=()):
```
I'm a little bit unclear; does this function get called every time we run the model, or only once when we cudagraphify

ezyang commented on torch/_inductor/compile_fx.py:
```diff
@@ -219,88 +225,649 @@ def remove_unaligned_input_idxs(inputs, static_input_idxs):
+        assert isinstance(inputs, (list, tuple))
+        self.graph = torch.cuda.CUDAGraph()
+        self.model = model
+        static_input_idxs = remove_unaligned_input_idxs(inputs, static_input_idxs)
```
I know this is old code, but it occurs to me we should probably warn when an unaligned parameter induces copies into cuda graphs

ezyang commented on torch/_inductor/compile_fx.py:
```diff
@@ -219,88 +225,649 @@ def remove_unaligned_input_idxs(inputs, static_input_idxs):
+
+        tape_manager = tape_manager_container.get_tape_manager(self)
+        tape_manager is not None
+        self.tape_manager = tape_manager
```
This also looks reference-cycle-y. What's the overall ownership structure for tapes and tape managers and cudagraphify objects?

ezyang commented on torch/_inductor/compile_fx.py:
```diff
@@ -219,88 +225,649 @@ def remove_unaligned_input_idxs(inputs, static_input_idxs):
+        self.recording_tape.increment_recording_tape(graph_id)
+
+        if len(self.head_to_tape_idx) < len(self.tapes):
+            self.head_to_tape_idx[graph_id] = self.active_recording_idx
```
not sure what's going on here

ezyang commented on torch/_inductor/compile_fx.py:
```diff
@@ -219,88 +225,649 @@ def remove_unaligned_input_idxs(inputs, static_input_idxs):
+        # if the live tensors count is set, then the memory pool is still
+        # being used even if the tape manager and CudaGraphTape objects are dead
+        if self.live_tensors_count:
+            return None
```
What does None return value mean here?

ezyang commented on torch/_inductor/compile_fx.py:
```diff
@@ -219,88 +225,649 @@ def remove_unaligned_input_idxs(inputs, static_input_idxs):
+        self.live_tensors_count = 0
+        # Used to keep alive the memory pool if the CudaGraph object dies
+        self.graph = None
+        self.lock = threading.Lock()
```
it suuure would be helpful to have types on these, especially since you seem to be monkeying around with this values from outside of the class directly

ezyang commented on torch/_inductor/compile_fx.py:
```diff
@@ -219,88 +225,649 @@ def remove_unaligned_input_idxs(inputs, static_input_idxs):
+            tape_manager = CUDAGraphTapeManger()
+            self.tape_manager = weakref.ref(tape_manager)
+        else:
+            tape_manager = self.tape_manager()
```
This is a race, the tape manager could die in between when you test it and when you use it.

ezyang commented on torch/_inductor/compile_fx.py:
```diff
@@ -219,88 +225,649 @@ def remove_unaligned_input_idxs(inputs, static_input_idxs):
+                self.live_tensors_count += 1
+                weakref.finalize(t, self.finalize_tensor)
+
+    def get_tape_manager(self, obj: Optional["CudaGraphify"] = None):
```
when do you pass None into the function here

ezyang commented on torch/_inductor/compile_fx.py:
```diff
@@ -219,88 +225,649 @@ def remove_unaligned_input_idxs(inputs, static_input_idxs):
+                self.alive_cudagraphify_objs += 1
+
+            # needs to be weak refs otherwise finalizer would keep obj alive
+            obj_ref = weakref.ref(obj)
```
but... you don't actually use obj_ref anywhere??

ezyang commented on torch/_inductor/compile_fx.py:
```diff
@@ -219,88 +225,649 @@ def remove_unaligned_input_idxs(inputs, static_input_idxs):
+                self.tape_manager = None
+                self.graph = None
+
+    def finalize_cuda_graphify(self, graph: torch.cuda.CUDAGraph, tape_manager):
```
What's the difference between `self.tape_manager` and the `tape_manager` argument here?

ezyang commented on torch/_inductor/compile_fx.py:
```diff
@@ -219,88 +225,649 @@ def remove_unaligned_input_idxs(inputs, static_input_idxs):
+        self.graph = None
+        self.lock = threading.Lock()
+
+    def finalize_tensor(self):
```
all of these methods could use docblocks...

ezyang commented on torch/_inductor/compile_fx.py:
```diff
@@ -219,88 +225,649 @@ def remove_unaligned_input_idxs(inputs, static_input_idxs):
+        static_input_idxs = remove_unaligned_input_idxs(inputs, static_input_idxs)
+
+        tape_manager = tape_manager_container.get_tape_manager(self)
+        tape_manager is not None
```
assert?

ezyang commented:
I need a more detailed PR description about the nuts and bolts inside the PR. Things I'm looking for:

* How does TapeManagerContainer work, what is the lifetime of this stuff
* Some examples of how execution proceeds for various edge cases
* What exactly is the tape representation
* What exactly is the tape lifecycle, esp re increment recording/executing tape
* What's up with liveness delta?
* What is the tape manager life cycle?

eellison committed:
Update on "cudagraph tapes'"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

This is a first PR implementing the following plan of POR:

- We will use a single Memory Pool across all cudagraphs. The memory pool will contain inputs (?), intermediates, and outputs. CUDAGraphs memory pool ensures that memory allocated within graph capture is not allocated outside of the memory pool, so that memory addresses remain static during graph replay. However, within the memory pool, normal storage allocation occurs, so there is no memory overhead. So long as we invoke separate CUDAGraphs in a consistent order as they are recorded, and do not retain extraneous references to Tensors, there should be no overhead. We will keep a global tape of the initial graph recordings and subsequent usages. Enforcing consistent CUDAGraph usage will also allow us to avoid copying over inputs to a CUDAGraph that are the outputs of a prior one. 
- To avoid “static outputs” being permanently allocated, which interferes with both memory being reclaimed within CUDAGraphs and [gradient stealing](https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/functions/accumulate_grad.h#L52), we will reconstruct output tensors Tensors with the appropriate fixed storage data pointers and other metadata on each cudagraph invocation. 
- When a CUDAGraph usage diverges from the existing tape we can 1. free the memory pool to the rest of the cuda allocator, invalidating the CUDAGraphs of our tape but keeping the existing Tensors alive or 2. Run torch inductor without cudagraphs for the remainder of the tape. The latter would have some memory penalty because it would not use memory from the memory pool 
- Because cudagraphs freeze memory addresses, the invocation of a cudagraph will replace whatever values the memory had previously. To avoid overwriting the live output values of a previous invocation, we can either check that the backward has been run for a cudagraph forward (not clear to me this is sufficient ?). We potentially could also do weak-ref ref counting.

Not everything is implemented yet, and I will not land this pr without subsequent follow ups. 

Follows ups, for correctness/minimal viable product in terms of priority include :
- Registering the cudagraphs as thread local state that needs to be copied over in autograd
- Ensuring clean up works successfully
- Adding tests for all the edge cases provided reviewers are okay with approach
- Handling subsequent usage when a graph invariant is broken, i.e. moving cuda graph allocations from the private memory pool to eager

And then further work would be:
- Allowing multiple, separate cudagraph paths by making the cuda caching allocator resumable and doing new recordings. 
- Adding configs to weaken some of the correctness conditions (for power user who can verify correctness)



[ghstack-poisoned]

eellison committed:
Update on "cudagraph tapes'"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

This is a first PR implementing the following plan of POR:

- We will use a single Memory Pool across all cudagraphs. The memory pool will contain inputs (?), intermediates, and outputs. CUDAGraphs memory pool ensures that memory allocated within graph capture is not allocated outside of the memory pool, so that memory addresses remain static during graph replay. However, within the memory pool, normal storage allocation occurs, so there is no memory overhead. So long as we invoke separate CUDAGraphs in a consistent order as they are recorded, and do not retain extraneous references to Tensors, there should be no overhead. We will keep a global tape of the initial graph recordings and subsequent usages. Enforcing consistent CUDAGraph usage will also allow us to avoid copying over inputs to a CUDAGraph that are the outputs of a prior one. 
- To avoid “static outputs” being permanently allocated, which interferes with both memory being reclaimed within CUDAGraphs and [gradient stealing](https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/functions/accumulate_grad.h#L52), we will reconstruct output tensors Tensors with the appropriate fixed storage data pointers and other metadata on each cudagraph invocation. 
- When a CUDAGraph usage diverges from the existing tape we can 1. free the memory pool to the rest of the cuda allocator, invalidating the CUDAGraphs of our tape but keeping the existing Tensors alive or 2. Run torch inductor without cudagraphs for the remainder of the tape. The latter would have some memory penalty because it would not use memory from the memory pool 
- Because cudagraphs freeze memory addresses, the invocation of a cudagraph will replace whatever values the memory had previously. To avoid overwriting the live output values of a previous invocation, we can either check that the backward has been run for a cudagraph forward (not clear to me this is sufficient ?). We potentially could also do weak-ref ref counting.

Not everything is implemented yet, and I will not land this pr without subsequent follow ups. 

Follows ups, for correctness/minimal viable product in terms of priority include :
- Registering the cudagraphs as thread local state that needs to be copied over in autograd
- Ensuring clean up works successfully
- Adding tests for all the edge cases provided reviewers are okay with approach
- Handling subsequent usage when a graph invariant is broken, i.e. moving cuda graph allocations from the private memory pool to eager

And then further work would be:
- Allowing multiple, separate cudagraph paths by making the cuda caching allocator resumable and doing new recordings. 
- Adding configs to weaken some of the correctness conditions (for power user who can verify correctness)



[ghstack-poisoned]

eellison committed:
Update on "cudagraph tapes'"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

This is a first PR implementing the following plan of POR:

- We will use a single Memory Pool across all cudagraphs. The memory pool will contain inputs (?), intermediates, and outputs. CUDAGraphs memory pool ensures that memory allocated within graph capture is not allocated outside of the memory pool, so that memory addresses remain static during graph replay. However, within the memory pool, normal storage allocation occurs, so there is no memory overhead. So long as we invoke separate CUDAGraphs in a consistent order as they are recorded, and do not retain extraneous references to Tensors, there should be no overhead. We will keep a global tape of the initial graph recordings and subsequent usages. Enforcing consistent CUDAGraph usage will also allow us to avoid copying over inputs to a CUDAGraph that are the outputs of a prior one. 
- To avoid “static outputs” being permanently allocated, which interferes with both memory being reclaimed within CUDAGraphs and [gradient stealing](https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/functions/accumulate_grad.h#L52), we will reconstruct output tensors Tensors with the appropriate fixed storage data pointers and other metadata on each cudagraph invocation. 
- When a CUDAGraph usage diverges from the existing tape we can 1. free the memory pool to the rest of the cuda allocator, invalidating the CUDAGraphs of our tape but keeping the existing Tensors alive or 2. Run torch inductor without cudagraphs for the remainder of the tape. The latter would have some memory penalty because it would not use memory from the memory pool 
- Because cudagraphs freeze memory addresses, the invocation of a cudagraph will replace whatever values the memory had previously. To avoid overwriting the live output values of a previous invocation, we can either check that the backward has been run for a cudagraph forward (not clear to me this is sufficient ?). We potentially could also do weak-ref ref counting.

Not everything is implemented yet, and I will not land this pr without subsequent follow ups. 

Follows ups, for correctness/minimal viable product in terms of priority include :
- Registering the cudagraphs as thread local state that needs to be copied over in autograd
- Ensuring clean up works successfully
- Adding tests for all the edge cases provided reviewers are okay with approach
- Handling subsequent usage when a graph invariant is broken, i.e. moving cuda graph allocations from the private memory pool to eager

And then further work would be:
- Allowing multiple, separate cudagraph paths by making the cuda caching allocator resumable and doing new recordings. 
- Adding configs to weaken some of the correctness conditions (for power user who can verify correctness)



[ghstack-poisoned]

eellison committed:
Update on "cudagraph trees"


CUDA Graph Trees 

Design doc: https://docs.google.com/document/d/1ZrxLGWz7T45MSX6gPsL6Ln4t0eZCSfWewtJ_qLd_D0E/edit

Not currently implemented :

- Right now, we are using weak tensor refs from outputs to check if a tensor has dies. This doesn't work because a) aliasing, and b) aot_autograd detaches tensors (see note [Detaching saved tensors in AOTAutograd]). Would neither https://github.com/pytorch/pytorch/issues/91395 to land to use storage weak refs or manually add a deleter fn that does what I want. This is doable but theres some interactions with the caching allocator checkpointing so saving for a stacked pr.

- Reclaiming memory from the inputs during model recording. This isn't terribly difficult but deferring to another PR. You would need to write over the input memory during warmup, and therefore copy the inputs to cpu. Saving for a stacked pr.



[ghstack-poisoned]

eellison committed:
Update on "cudagraph trees"


CUDA Graph Trees 

Design doc: https://docs.google.com/document/d/1ZrxLGWz7T45MSX6gPsL6Ln4t0eZCSfWewtJ_qLd_D0E/edit

Not currently implemented :

- Right now, we are using weak tensor refs from outputs to check if a tensor has dies. This doesn't work because a) aliasing, and b) aot_autograd detaches tensors (see note [Detaching saved tensors in AOTAutograd]). Would need either https://github.com/pytorch/pytorch/issues/91395 to land to use storage weak refs or manually add a deleter fn that does what I want. This is doable but theres some interactions with the caching allocator checkpointing so saving for a stacked pr.

- Reclaiming memory from the inputs during model recording. This isn't terribly difficult but deferring to another PR. You would need to write over the input memory during warmup, and therefore copy the inputs to cpu. Saving for a stacked pr.

- Warning on overwriting previous generation outputs. and handling nested torch.compile() calls in generation tracking


[ghstack-poisoned]

eellison committed:
Update on "cudagraph trees"


CUDA Graph Trees 

Design doc: https://docs.google.com/document/d/1ZrxLGWz7T45MSX6gPsL6Ln4t0eZCSfWewtJ_qLd_D0E/edit

Not currently implemented :

- Right now, we are using weak tensor refs from outputs to check if a tensor has dies. This doesn't work because a) aliasing, and b) aot_autograd detaches tensors (see note [Detaching saved tensors in AOTAutograd]). Would need either https://github.com/pytorch/pytorch/issues/91395 to land to use storage weak refs or manually add a deleter fn that does what I want. This is doable but theres some interactions with the caching allocator checkpointing so saving for a stacked pr.

- Reclaiming memory from the inputs during model recording. This isn't terribly difficult but deferring to another PR. You would need to write over the input memory during warmup, and therefore copy the inputs to cpu. Saving for a stacked pr.

- Warning on overwriting previous generation outputs. and handling nested torch.compile() calls in generation tracking


[ghstack-poisoned]

eellison committed:
Update on "cudagraph trees"


CUDA Graph Trees 

Design doc: https://docs.google.com/document/d/1ZrxLGWz7T45MSX6gPsL6Ln4t0eZCSfWewtJ_qLd_D0E/edit

Not currently implemented :

- Right now, we are using weak tensor refs from outputs to check if a tensor has dies. This doesn't work because a) aliasing, and b) aot_autograd detaches tensors (see note [Detaching saved tensors in AOTAutograd]). Would need either https://github.com/pytorch/pytorch/issues/91395 to land to use storage weak refs or manually add a deleter fn that does what I want. This is doable but theres some interactions with the caching allocator checkpointing so saving for a stacked pr.

- Reclaiming memory from the inputs during model recording. This isn't terribly difficult but deferring to another PR. You would need to write over the input memory during warmup, and therefore copy the inputs to cpu. Saving for a stacked pr.

- Warning on overwriting previous generation outputs. and handling nested torch.compile() calls in generation tracking


[ghstack-poisoned]

ezyang commented:
![Untitled-2023-03-03-1439](https://user-images.githubusercontent.com/13564/222815509-374f3400-f83d-4f7d-8fa6-4a092b3250bb.png)

ezyang commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -0,0 +1,863 @@
+    )
+
+
+class CUDAGraphTreeManager(object):
```
nit: don't need to explicitly subclass from object

ezyang commented on torch/_inductor/compile_fx.py:
```diff
@@ -252,18 +252,26 @@ def run(new_inputs):
+        cudagraphify_impl as new_cudagraphify_impl,
+    )
+
+    if config.triton.cudagraph_trees:
```
Is it possible to just switch entirely to the new implementation, without the knob here?

eellison replied:
Not within this PR, after storages pr we could

ezyang commented on torch/_inductor/compile_fx.py:
```diff
@@ -282,27 +290,28 @@ def remove_unaligned_input_idxs(inputs, static_input_idxs):
     assert isinstance(inputs, (list, tuple))
     static_inputs = [
-        static_input(x) if idx not in static_input_idxs else x.detach()
+        static_input(x).zero_() if idx not in static_input_idxs else x.detach()
```
How come the zero_?

eellison replied:
I changed static_input to no longer zero_ since i'm calling it during recording, so now we need to

ezyang commented on torch/_inductor/config.py:
```diff
@@ -165,6 +165,11 @@ class triton:
+    # Use cudagraph trees for memory pooling if `cudagraphs` is True
+    cudagraph_trees = True
+
+    debug_cudagraph_trees = True
```
True by default?

eellison replied:
it's not doing any logging, it's just adding runtime checks, which I think we should have on initiially

ezyang commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -0,0 +1,863 @@
+                self.graph = None
+
+                # manager was used again after existing cleanup,
+                # we shouldnt set it to None
```
"was" or "will be"?

eellison replied:
was

ezyang commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -0,0 +1,863 @@
+
+        weakref.finalize(fn, self.finalize_reference)
+
+    def finalize_tree_manager(self):
```
It would be extremely helpful if methods like this which are never called externally have a leading underscore, showing they're class private.

ezyang commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -0,0 +1,863 @@
+    """
+    Manages the lifetime of the tree manager. Like `PrivatePool` in cuda caching allocator,
+    the tree and its corresponding memory pool should be kept alive as long as any outstanding
+    graph or tensor which is an output of a graph remains alive.
```
An explanation of the lifecycle of this object would be very useful. Based on reading, it seems like it is something like:

1. Is constructed, no graph, no fns, no tensors
2. Tree manager is fetched, resulting in tree manager being allocated
3. We generate a bunch of functions, calling add_strong_reference
4. These functions die, calling finalize_reference
5. When all the functions die, we finalize_tree_manager. We look for all the live tensors and add references to THOSE
6. We count as tensors die
7. All the tensors are dead, we deallocate the tree manager

Something that is not immediately obvious from reading this class only is the step 2/3 of the life cycle: what if you start building the tree manager but the first fn you make strong reference to dies. Won't you die prematurely? What if I then try to add another strong reference after having already finalized?

eellison replied:
> what if you start building the tree manager but the first fn you make strong reference to dies. Won't you die prematurely?

If the reference is dead, how is it dying prematurely. That is dying at the right time, no ? We only die when all references to the manager is gone. If there are multiple references and only one refernece dies, it wont be cleaned up

eellison replied:
Added this comment inline

ezyang commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -0,0 +1,863 @@
+    In order to support recording a subsequent cuda graph recording after execution of this graph,
+    we checkpoint the state of the memory pool so that it may later be resumed.
+
+    See [setCheckpointPoolState] for further explanation.
```
Link to the diagram too here?

eellison replied:
By diagram, do you mean the ascii diagram or the diagram that we made on excaldiraw ?

ezyang replied:
excalidraw

ezyang commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -0,0 +1,863 @@
+        self.previous_recording_outputs: List[TensorWeakRef] = []
+
+        self.debug_fail_counter = 0
+        self.debug_checkpointing_counter = 0
```
Would appreciate comments on all of these properties

ezyang replied:
I particularly would like to know what "roots" and "previous_recording_outputs" are.

eellison replied:
`previous_recording_outputs` will be simplified in next pr with storages, so let's not overly focus on it here. But yess will add comments

ezyang commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -0,0 +1,863 @@
+        """
+        assert self.in_recording
+
+        if self.current_node is None:
```
When can this happen?

eellison replied:
Added comment:
```
        # the most recently invoked cudagraph wrapping of a function. Will be None 
        # when there is no output from a previous recording or execution whose memory 
        # we need to respect in the cuda caching allocator, which includes if that 
        # memory is from a separate generation
```

ezyang commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -0,0 +1,863 @@
+                    return self.execute_node(child, new_inputs)
+
+            # now that we know the new function can't be run as a child of the
+            # current node, if it is a root, try to end the current execution.
```
So, suppose I have something like A - B - C - B - C - B - ... I will just keep tacking things onto the path, won't I?

eellison replied:
No, because of the generation thing

ezyang commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -0,0 +1,863 @@
+
+            self.try_end_curr_execution()
+            if self.current_node is not None:
+                self.convert_execution_to_recording()
```
dumb Q: where did the execution here happen?

eellison replied:
I renamed the function, does that clarify things?

ezyang commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -0,0 +1,863 @@
+    """
+
+    def __init__(self):
+        self.roots: Dict[FunctionID, List[CUDAGraphNode]] = defaultdict(list)
```
Is there a bound to how many CUDAGraphNodes in this list could there be?

eellison replied:
No bound, although you would expect the invariants to not change much

ezyang replied:
I guess my question is, can we get into some sort of horrible situation where we just keep allocating new nodes without deallocating them.

eellison replied:
If you are in a sufficiently dynamic model there's not really a good way to avoid that. Could warn after some threshold. But as of now we're not always enabling it.

ezyang commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -0,0 +1,863 @@
+        self.current_node.clear_path_outputs()
+        self.current_node = None
+
+    def convert_execution_to_recording(self):
```
So there's some invariant this has to be called... after `record_function`, I guess??

ezyang replied:
Actually, I think I misread this function name: this is taking us from "execution" mode and putting us in "recording" mode, right? "convert" is very misleading as I think of some sort of data structure conversion in that case.

ezyang replied:
This function really can use the comment.

eellison replied:
Added a comment and renamed. let me know if you think it needs further clarification

eellison commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -0,0 +1,863 @@
+                # manager was used again after existing cleanup,
+                # we shouldnt set it to None
+                if self.live_cudagraphify_fns == 0:
+                    self.tree_manager = None
```
That's not handled. If so, they should call `add_strong_reference`

eellison committed:
Update on "cudagraph trees"


CUDA Graph Trees 

Design doc: https://docs.google.com/document/d/1ZrxLGWz7T45MSX6gPsL6Ln4t0eZCSfWewtJ_qLd_D0E/edit

Not currently implemented :

- Right now, we are using weak tensor refs from outputs to check if a tensor has dies. This doesn't work because a) aliasing, and b) aot_autograd detaches tensors (see note [Detaching saved tensors in AOTAutograd]). Would need either https://github.com/pytorch/pytorch/issues/91395 to land to use storage weak refs or manually add a deleter fn that does what I want. This is doable but theres some interactions with the caching allocator checkpointing so saving for a stacked pr.

- Reclaiming memory from the inputs during model recording. This isn't terribly difficult but deferring to another PR. You would need to write over the input memory during warmup, and therefore copy the inputs to cpu. Saving for a stacked pr.

- Warning on overwriting previous generation outputs. and handling nested torch.compile() calls in generation tracking


[ghstack-poisoned]

eellison committed:
Update on "cudagraph trees"


CUDA Graph Trees 

Design doc: https://docs.google.com/document/d/1ZrxLGWz7T45MSX6gPsL6Ln4t0eZCSfWewtJ_qLd_D0E/edit

Not currently implemented :

- Right now, we are using weak tensor refs from outputs to check if a tensor has dies. This doesn't work because a) aliasing, and b) aot_autograd detaches tensors (see note [Detaching saved tensors in AOTAutograd]). Would need either https://github.com/pytorch/pytorch/issues/91395 to land to use storage weak refs or manually add a deleter fn that does what I want. This is doable but theres some interactions with the caching allocator checkpointing so saving for a stacked pr.

- Reclaiming memory from the inputs during model recording. This isn't terribly difficult but deferring to another PR. You would need to write over the input memory during warmup, and therefore copy the inputs to cpu. Saving for a stacked pr.

- Warning on overwriting previous generation outputs. and handling nested torch.compile() calls in generation tracking


[ghstack-poisoned]

ezyang commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -0,0 +1,904 @@
+        # the most recently invoked cudagraph wrapping of a function. Will be None
+        # when there is no output from a previous recording or execution whose memory
+        # we need to respect in the cuda caching allocator, which includes if that
+        # memory is from a separate generation.
```
after comma, maybe clearer to say "if you incremented generation, this will also be None?"

ezyang commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -0,0 +1,904 @@
+        self.current_node.clear_path_outputs()
+        self.current_node = None
+
+    def checkpoint_execution_state_in_allocator(self):
```
to be clear, this isn't *saving* a checkpoint: it's applying one, right?

ezyang commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -0,0 +1,863 @@
+        device = self.current_node.device
+        assert state is not None and device is not None
+
+        stale_tensors = [t() for t in self.previous_recording_outputs if is_live(t)]
```
this is the "will be removed soon" right. Maybe should'a squashed the PRs 🤔

eellison replied:
Yea, I can squash the prs.

ezyang commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -0,0 +1,904 @@
+
+    def try_end_curr_execution(self) -> None:
+        """
+        Check if the current executng node can be terminated, either because all outputs of the
```
executing

ezyang commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -0,0 +1,904 @@
+        # in children to avoid children having to chase parent pointers in the hot path
+        # DO NOT reassign output_weakrefs, only call `clear()`
+        self.outputs_weakrefs: List[Optional[TensorWeakRef]] = []
+        self.path_weakrefs: List[List[Optional[TensorWeakRef]]] = [
```
I don't actually know what a "path" is lol

eellison replied:
Path from ancestor to root. will add comment

eellison replied:
It's graph I output J, but yea, will edit

eellison committed:
Update on "cudagraph trees"


CUDA Graph Trees 

Design doc: https://docs.google.com/document/d/1ZrxLGWz7T45MSX6gPsL6Ln4t0eZCSfWewtJ_qLd_D0E/edit

Not currently implemented :

- Right now, we are using weak tensor refs from outputs to check if a tensor has dies. This doesn't work because a) aliasing, and b) aot_autograd detaches tensors (see note [Detaching saved tensors in AOTAutograd]). Would need either https://github.com/pytorch/pytorch/issues/91395 to land to use storage weak refs or manually add a deleter fn that does what I want. This is doable but theres some interactions with the caching allocator checkpointing so saving for a stacked pr.

- Reclaiming memory from the inputs during model recording. This isn't terribly difficult but deferring to another PR. You would need to write over the input memory during warmup, and therefore copy the inputs to cpu. Saving for a stacked pr.

- Warning on overwriting previous generation outputs. and handling nested torch.compile() calls in generation tracking


[ghstack-poisoned]

eellison commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -0,0 +1,904 @@
+        # when they are first invoked, none of their inputs are outputs are outputs
+        # of another node, nor are there any live outputs of another node whose
+        # liveness would create a dependency.
+        self.roots: Dict[FunctionID, List[CUDAGraphNode]] = defaultdict(list)
```
I did call it cuda graph tree **s**  lol. imo, forest is a much less well known term that Tree. Do you have a strong preference for one or the other ?

eellison committed:
Update on "cudagraph trees"


CUDA Graph Trees 

Design doc: https://docs.google.com/document/d/1ZrxLGWz7T45MSX6gPsL6Ln4t0eZCSfWewtJ_qLd_D0E/edit

Not currently implemented :

- Right now, we are using weak tensor refs from outputs to check if a tensor has dies. This doesn't work because a) aliasing, and b) aot_autograd detaches tensors (see note [Detaching saved tensors in AOTAutograd]). Would need either https://github.com/pytorch/pytorch/issues/91395 to land to use storage weak refs or manually add a deleter fn that does what I want. This is doable but theres some interactions with the caching allocator checkpointing so saving for a stacked pr.

- Reclaiming memory from the inputs during model recording. This isn't terribly difficult but deferring to another PR. You would need to write over the input memory during warmup, and therefore copy the inputs to cpu. Saving for a stacked pr.

- Warning on overwriting previous generation outputs. and handling nested torch.compile() calls in generation tracking


[ghstack-poisoned]

ngimel commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -0,0 +1,906 @@
+        # needed to reconstruct instead.
+        for out in self.recording_outputs:
+            if isinstance(out, torch.Tensor):
+                self.device = (
```
how would that work if function spans multiple devices? Should you assert that `out.device.index` is the same as `self.device`?

eellison replied:
I limit the cudagraphs with new api to a single device in a follow up pr, in the warm up one

ngimel commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -0,0 +1,906 @@
+        assert len(self.outputs_weakrefs) == 0
+        weak_refs = [self._map_to_ref(o) for o in outputs]
+        self.outputs_weakrefs.extend(weak_refs)
+        self.recorded_liveness_after_graph = self._get_liveness(self.path_weakrefs)
```
did `self.path_weakrefs` change between here and line 419?

eellison replied:
Yes, `self.outputs_weakrefs.extend(weak_refs)` would have changed it. i'll add comment

eellison committed:
Update on "cudagraph trees"


CUDA Graph Trees 

Design doc: https://docs.google.com/document/d/1ZrxLGWz7T45MSX6gPsL6Ln4t0eZCSfWewtJ_qLd_D0E/edit

Not currently implemented :

- Right now, we are using weak tensor refs from outputs to check if a tensor has dies. This doesn't work because a) aliasing, and b) aot_autograd detaches tensors (see note [Detaching saved tensors in AOTAutograd]). Would need either https://github.com/pytorch/pytorch/issues/91395 to land to use storage weak refs or manually add a deleter fn that does what I want. This is doable but theres some interactions with the caching allocator checkpointing so saving for a stacked pr.

- Reclaiming memory from the inputs during model recording. This isn't terribly difficult but deferring to another PR. You would need to write over the input memory during warmup, and therefore copy the inputs to cpu. Saving for a stacked pr.

- Warning on overwriting previous generation outputs. and handling nested torch.compile() calls in generation tracking


[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D43999887).

ezyang commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -0,0 +1,906 @@
+        self.id = id
+
+        # if this is a root parent will be None. use weakref to prevent reference cycle
+        self._parent = weakref.ref(parent) if parent is not None else None
```
@eellison, one of the things I really like about reading your PRs is I know you will be very good about handling ref cycles haha :D

ezyang commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -0,0 +1,906 @@
+        id: GraphID,
+        parent: Optional[CUDAGraphNode],
+        inputs: List[Tensor],
+        cuda_graphs_pool: Tuple[int, int],
```
What's this

ngimel replied:
lol I was surprised by that too, it's an `Opaque token (returned by a call to [graph_pool_handle()] or [other_Graph_instance.pool()])` according to the doc, but it looks like an `(int, int)`

eellison replied:
This is the memory pool handle to use a single pool across separate recordings

ezyang commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -0,0 +1,906 @@
+        )
+
+        # we preserve a single reference to executed outputs that is then referenced
+        # in children to avoid children having to chase parent pointers in the hot path
```
There seems to be some invariant here, it would be nice if it was explicitly written out

ezyang replied:
In particular, according to the comment, it should be possible to recompute the output weakrefs by chasing parent pointers. We should implement this and have a debug mode to actually check that this is consistent.

eellison replied:
Can implement this. I dont think there is any invariant here besides what is commented.

ezyang commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -0,0 +1,906 @@
+        # the cudagraphify_fns. Reference to the Graph is needed to keep the private pool from
+        # deallocation.
+        self.live_tensors_count = 0
+        self.graph: Optional[torch.cuda.CUDAGraph] = None
```
Not sure what this is

eellison replied:
"Reference to the Graph is needed to keep the private pool from deallocation." 

All the cudagraphify_fns are dead, but we keep graph around to prevent deallocation. i simplify this in a later PR.. :'(

ezyang commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -0,0 +1,906 @@
+        self.expected_dead_indices_after_graph: List[Tuple[int, int]] = []
+
+        # all live indices after graph recording
+        self.live_indices_after_graph: List[Tuple[int, int]] = []
```
Also level i, output j?

ezyang commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -0,0 +1,906 @@
+        self.recorded_liveness_before_graph: List[List[bool]] = []
+        self.recorded_liveness_after_graph: List[List[bool]] = []
+
+        # indices into node at level i for output j
```
I have no idea, at thisnpoint, what a "level" is

eellison replied:
lol took that terminology from wikipedia. the level of the tree is distance from the root. maybe depth is more idiomatic. or i'll just spell out with the full description..

ezyang replied:
I agree level is unambiguous for this. It's surprising though because you don't have a root, you would traverse parent pointers to populate this.

eellison replied:
It might be cleaner for me to pass in a root, instead of a parent. it doesn't matter perf wise (dont traverse on hot path) but it would be cleaner perhaps

ezyang commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -0,0 +1,906 @@
+        self.live_indices_after_graph: List[Tuple[int, int]] = []
+
+        if self.parent is not None:
+            previous_liveness = self.parent.recorded_liveness_after_graph
```
Paradoxically, I think the names here are a bit too long, and shorter names would have helped me for clarity here

eellison replied:
what do you think would be better ?

ezyang commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -0,0 +1,906 @@
+
+        if self.parent is not None:
+            previous_liveness = self.parent.recorded_liveness_after_graph
+            curr_liveness = self._get_liveness(self.path_weakrefs)
```
If these methods don't actually depend on entries in self, I might suggest declaring them as top level functions to make it clear they don't access self in an unrestricted way.

ezyang commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -0,0 +1,906 @@
+            previous_liveness = self.parent.recorded_liveness_after_graph
+            curr_liveness = self._get_liveness(self.path_weakrefs)
+
+            different_indices = self._get_different_indices(
```
A small data structure for indices might help with code organization here

ezyang commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -0,0 +1,906 @@
+
+        # graph needs to be kept alive for the duration of the model
+        # otherwise the memory pool would be freed on the first recording
+        inps_alloc_graph = torch.cuda.CUDAGraph()
```
Needing two cuda graphs is surprising

eellison replied:
this is cleaned up in a later pr

eellison committed:
Update on "cudagraph trees"


CUDA Graph Trees 

Design doc: https://docs.google.com/document/d/1ZrxLGWz7T45MSX6gPsL6Ln4t0eZCSfWewtJ_qLd_D0E/edit

Not currently implemented :

- Right now, we are using weak tensor refs from outputs to check if a tensor has dies. This doesn't work because a) aliasing, and b) aot_autograd detaches tensors (see note [Detaching saved tensors in AOTAutograd]). Would need either https://github.com/pytorch/pytorch/issues/91395 to land to use storage weak refs or manually add a deleter fn that does what I want. This is doable but theres some interactions with the caching allocator checkpointing so saving for a stacked pr.

- Reclaiming memory from the inputs during model recording. This isn't terribly difficult but deferring to another PR. You would need to write over the input memory during warmup, and therefore copy the inputs to cpu. Saving for a stacked pr.

- Warning on overwriting previous generation outputs. and handling nested torch.compile() calls in generation tracking

Differential Revision: [D43999887](https://our.internmc.facebook.com/intern/diff/D43999887)

[ghstack-poisoned]

eellison committed:
Update on "cudagraph trees"


CUDA Graph Trees 

Design doc: https://docs.google.com/document/d/1ZrxLGWz7T45MSX6gPsL6Ln4t0eZCSfWewtJ_qLd_D0E/edit

Not currently implemented :

- Right now, we are using weak tensor refs from outputs to check if a tensor has dies. This doesn't work because a) aliasing, and b) aot_autograd detaches tensors (see note [Detaching saved tensors in AOTAutograd]). Would need either https://github.com/pytorch/pytorch/issues/91395 to land to use storage weak refs or manually add a deleter fn that does what I want. This is doable but theres some interactions with the caching allocator checkpointing so saving for a stacked pr.

- Reclaiming memory from the inputs during model recording. This isn't terribly difficult but deferring to another PR. You would need to write over the input memory during warmup, and therefore copy the inputs to cpu. Saving for a stacked pr.

- Warning on overwriting previous generation outputs. and handling nested torch.compile() calls in generation tracking

Differential Revision: [D43999887](https://our.internmc.facebook.com/intern/diff/D43999887)

[ghstack-poisoned]

eellison commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -0,0 +1,906 @@
+            idx
+            for idx, t in enumerate(inputs)
+            if self._is_cuda_graph_recorded_tensor(t)
+        ]
```
There's no way currently for there to be separate pool uses (nor would you really want there to be)

ezyang commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -0,0 +1,906 @@
+
+        self.device: int = next(
+            (x.device.index for x in inputs if is_cuda_tensor), None
+        )
```
I'm OK with squashing all of the PRs past this.

eellison commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -0,0 +1,906 @@
+
+        # For each node in path, which outputs were observed to be live
+        # before invoking graph recording, and after graph recording
+        self.recorded_liveness_before_graph: List[List[bool]] = []
```
added an invariant checker, and used type alias with comment

eellison committed:
Update on "cudagraph trees"


CUDA Graph Trees 

Design doc: https://docs.google.com/document/d/1ZrxLGWz7T45MSX6gPsL6Ln4t0eZCSfWewtJ_qLd_D0E/edit

Not currently implemented :

- Right now, we are using weak tensor refs from outputs to check if a tensor has dies. This doesn't work because a) aliasing, and b) aot_autograd detaches tensors (see note [Detaching saved tensors in AOTAutograd]). Would need either https://github.com/pytorch/pytorch/issues/91395 to land to use storage weak refs or manually add a deleter fn that does what I want. This is doable but theres some interactions with the caching allocator checkpointing so saving for a stacked pr.

- Reclaiming memory from the inputs during model recording. This isn't terribly difficult but deferring to another PR. You would need to write over the input memory during warmup, and therefore copy the inputs to cpu. Saving for a stacked pr.

- Warning on overwriting previous generation outputs. and handling nested torch.compile() calls in generation tracking

Differential Revision: [D43999887](https://our.internmc.facebook.com/intern/diff/D43999887)

[ghstack-poisoned]

jansel commented on test/inductor/test_torchinductor.py:
```diff
@@ -7576,6 +7578,358 @@ def test_rnn_compile_safe(self):
+    def all_live_block_count():
+        return len(all_live_blocks())
+
+    class CudaGraphTapeTests(TestCase):
```
Let's create a new file for these tests.  `test_torchinductor.py` is already too big.

eellison replied:
Will do as another pr in the stack, and land all at once, to minimize churn to other prs

ezyang replied:
I think you should squash all the PRs in the stack

jansel commented on torch/_inductor/compile_fx.py:
```diff
@@ -256,18 +256,26 @@ def run(new_inputs):
 
 @dynamo_utils.dynamo_timed
 def cudagraphify(model, inputs, static_input_idxs=()):
+    from torch._inductor.cudagraph_trees import (
```
Perhaps move all the cudagraphs related stuff (including this function) to the same file so we don't need to workaround circular imports like this.

eellison committed:
Update on "cudagraph trees"


CUDA Graph Trees 

Design doc: https://docs.google.com/document/d/1ZrxLGWz7T45MSX6gPsL6Ln4t0eZCSfWewtJ_qLd_D0E/edit

Not currently implemented :

- Right now, we are using weak tensor refs from outputs to check if a tensor has dies. This doesn't work because a) aliasing, and b) aot_autograd detaches tensors (see note [Detaching saved tensors in AOTAutograd]). Would need either https://github.com/pytorch/pytorch/issues/91395 to land to use storage weak refs or manually add a deleter fn that does what I want. This is doable but theres some interactions with the caching allocator checkpointing so saving for a stacked pr.

- Reclaiming memory from the inputs during model recording. This isn't terribly difficult but deferring to another PR. You would need to write over the input memory during warmup, and therefore copy the inputs to cpu. Saving for a stacked pr.

- Warning on overwriting previous generation outputs. and handling nested torch.compile() calls in generation tracking

Differential Revision: [D43999887](https://our.internmc.facebook.com/intern/diff/D43999887)

[ghstack-poisoned]

eellison committed:
Update on "cudagraph trees"


CUDA Graph Trees 

Design doc: https://docs.google.com/document/d/1ZrxLGWz7T45MSX6gPsL6Ln4t0eZCSfWewtJ_qLd_D0E/edit

Not currently implemented :

- Right now, we are using weak tensor refs from outputs to check if a tensor has dies. This doesn't work because a) aliasing, and b) aot_autograd detaches tensors (see note [Detaching saved tensors in AOTAutograd]). Would need either https://github.com/pytorch/pytorch/issues/91395 to land to use storage weak refs or manually add a deleter fn that does what I want. This is doable but theres some interactions with the caching allocator checkpointing so saving for a stacked pr.

- Reclaiming memory from the inputs during model recording. This isn't terribly difficult but deferring to another PR. You would need to write over the input memory during warmup, and therefore copy the inputs to cpu. Saving for a stacked pr.

- Warning on overwriting previous generation outputs. and handling nested torch.compile() calls in generation tracking

Differential Revision: [D43999887](https://our.internmc.facebook.com/intern/diff/D43999887)

[ghstack-poisoned]

eellison committed:
Update on "cudagraph trees"


CUDA Graph Trees 

Design doc: https://docs.google.com/document/d/1ZrxLGWz7T45MSX6gPsL6Ln4t0eZCSfWewtJ_qLd_D0E/edit

Not currently implemented :

- Right now, we are using weak tensor refs from outputs to check if a tensor has dies. This doesn't work because a) aliasing, and b) aot_autograd detaches tensors (see note [Detaching saved tensors in AOTAutograd]). Would need either https://github.com/pytorch/pytorch/issues/91395 to land to use storage weak refs or manually add a deleter fn that does what I want. This is doable but theres some interactions with the caching allocator checkpointing so saving for a stacked pr.

- Reclaiming memory from the inputs during model recording. This isn't terribly difficult but deferring to another PR. You would need to write over the input memory during warmup, and therefore copy the inputs to cpu. Saving for a stacked pr.

- Warning on overwriting previous generation outputs. and handling nested torch.compile() calls in generation tracking

Differential Revision: [D43999887](https://our.internmc.facebook.com/intern/diff/D43999887)

[ghstack-poisoned]

eellison committed:
Update on "cudagraph trees"


CUDA Graph Trees 

Design doc: https://docs.google.com/document/d/1ZrxLGWz7T45MSX6gPsL6Ln4t0eZCSfWewtJ_qLd_D0E/edit

Not currently implemented :

- Right now, we are using weak tensor refs from outputs to check if a tensor has dies. This doesn't work because a) aliasing, and b) aot_autograd detaches tensors (see note [Detaching saved tensors in AOTAutograd]). Would need either https://github.com/pytorch/pytorch/issues/91395 to land to use storage weak refs or manually add a deleter fn that does what I want. This is doable but theres some interactions with the caching allocator checkpointing so saving for a stacked pr.

- Reclaiming memory from the inputs during model recording. This isn't terribly difficult but deferring to another PR. You would need to write over the input memory during warmup, and therefore copy the inputs to cpu. Saving for a stacked pr.

- Warning on overwriting previous generation outputs. and handling nested torch.compile() calls in generation tracking

Differential Revision: [D43999887](https://our.internmc.facebook.com/intern/diff/D43999887)

[ghstack-poisoned]

eellison committed:
Update on "cudagraph trees"


CUDA Graph Trees 

Design doc: https://docs.google.com/document/d/1ZrxLGWz7T45MSX6gPsL6Ln4t0eZCSfWewtJ_qLd_D0E/edit

Not currently implemented :

- Right now, we are using weak tensor refs from outputs to check if a tensor has dies. This doesn't work because a) aliasing, and b) aot_autograd detaches tensors (see note [Detaching saved tensors in AOTAutograd]). Would need either https://github.com/pytorch/pytorch/issues/91395 to land to use storage weak refs or manually add a deleter fn that does what I want. This is doable but theres some interactions with the caching allocator checkpointing so saving for a stacked pr.

- Reclaiming memory from the inputs during model recording. This isn't terribly difficult but deferring to another PR. You would need to write over the input memory during warmup, and therefore copy the inputs to cpu. Saving for a stacked pr.

- Warning on overwriting previous generation outputs. and handling nested torch.compile() calls in generation tracking

Differential Revision: [D43999887](https://our.internmc.facebook.com/intern/diff/D43999887)

[ghstack-poisoned]

eellison committed:
Update on "cudagraph trees"


CUDA Graph Trees 

Design doc: https://docs.google.com/document/d/1ZrxLGWz7T45MSX6gPsL6Ln4t0eZCSfWewtJ_qLd_D0E/edit

Not currently implemented :

- Right now, we are using weak tensor refs from outputs to check if a tensor has dies. This doesn't work because a) aliasing, and b) aot_autograd detaches tensors (see note [Detaching saved tensors in AOTAutograd]). Would need either https://github.com/pytorch/pytorch/issues/91395 to land to use storage weak refs or manually add a deleter fn that does what I want. This is doable but theres some interactions with the caching allocator checkpointing so saving for a stacked pr.

- Reclaiming memory from the inputs during model recording. This isn't terribly difficult but deferring to another PR. You would need to write over the input memory during warmup, and therefore copy the inputs to cpu. Saving for a stacked pr.

- Warning on overwriting previous generation outputs. and handling nested torch.compile() calls in generation tracking

Differential Revision: [D43999887](https://our.internmc.facebook.com/intern/diff/D43999887)

[ghstack-poisoned]

eellison commented:
@pytorchbot  merge
